### PR TITLE
Add validators for bboxes annotation files

### PR DESCRIPTION
**Pull request by sfmig**
Rebase after #27 

## Description

**What is this PR**

- [ ] Bug fix
- [x] Addition of a new feature
- [ ] Other

**Why is this PR needed?**
To validate input data files with bounding boxes annotations

**What does this PR do?**
- Adds an `annotations/validator` module to define classes for valid JSON files with bounding boxes annotations
    - We support two formats: COCO-JSON and VIA-JSON
    - We use JSON schemas (via the [jsonschema](https://github.com/python-jsonschema/jsonschema) package) to check the types of the fields defined in a JSON file.
    - I added an `annotations/json_schemas` module with the default schemas for COCO and VIA and a `utils` module with helper functions.
- Adds tests
    - I include a test to check the schemas that we ship include the minimum required data (`test_required_keys_in_COCO_schema`, `test_required_keys_in_VIA_schema`). This is because the validation against a schema only checks the type for the keys in the schema that appear also in the file. So for example, if for any reason the schemas that we ship become empty dictionaries, the validation of a file against those schema will still pass.

**Question**
I am tempted to do away with the "check file is JSON" check (via `_check_file_is_json`) - the only reason I have it now is because the error message we overwrite is slightly more clear, but to be honest I am not sure it is worth it.

Any other benefit to doing this that I may be missing?
    
## References

#18 

## How has this PR been tested?

Tests pass locally and in CI.

## Is this a breaking change?

No.

## Does this PR require an update to the documentation?

Not for now.

## Checklist:

- [x] The code has been tested locally
- [x] Tests have been added to cover all new functionality
- [ n/a ] The documentation has been updated to reflect any changes
- [x] The code has been formatted with [pre-commit](https://pre-commit.com/)

### CODE REVIEW

File: ethology/annotations/validators.py [hunk 1]
```diff
@@ -0,0 +1,192 @@
...
+        If the COCO JSON file is missing any of the required keys.
+
+    """
+
+    path: Path = field()
+    schema: dict = field(
+        default=_get_default_COCO_schema(),
+        init=False,
+        # init=False makes the attribute to be unconditionally initialized
+        # with the specified default
```
Comment (niksirbi) [inline ethology/annotations/validators.py:143]: I'm not sure what that comment means. Does it mean I cannot overwrite the schema passing a different dict?
In any case, this comment should be moved to the first occurrence of `init=False` in this file.

File: ethology/annotations/json_schemas/schemas/README.md [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+
+## VIA schema
+
+The VIA schema corresponds to the format exported by VGG Image Annotator 2.x.y (VIA) for object detection annotations.
+
+Each image under `_via_img_metadata` is indexed using a unique key: FILENAME-FILESIZE. We use "additionalProperties" to allow for any key name, see https://stackoverflow.com/a/69811612/24834957.
+
+The section `_via_image_id_list` contains an ordered list of image keys using a unique key: `FILENAME-FILESIZE`, the position in the list defines the image ID.
+
+The section `_via_attributes` region attributes and file attributes, to display in VIA's UI and to classify the data.
```
Comment (niksirbi) [inline ethology/annotations/json_schemas/schemas/README.md:17]: feels like a verb is missing here, "contains"?

File: tests/test_unit/test_annotations/test_validators.py [hunk 3]
```diff
@@ -0,0 +1,408 @@
...
+            pytest.raises(jsonschema.exceptions.ValidationError),
+            "3 is not of type 'object'",
+        ),
+    ],
+)
+def test_validators_invalid_input_files(
+    invalid_input_file: str,
+    validator: type[ValidVIA | ValidCOCO],
+    expected_exception: pytest.raises,
+    log_message: str,
```
Comment (niksirbi) [inline tests/test_unit/test_annotations/test_validators.py:193]: what you are actualling checking should be called "error_message" not "log_message", write. There is no logging setup yet, and you are not checking tha logs, just the exception info.

**Review by niksirbi — APPROVED** [review]
Excellent work as usual @sfmig!

I see nothing fundamentally wrong with your approach here, seems sensible to me.
The functionality is also sufficiently covered by tests imo.

I've left a few specific comments/questions/suggestions, but nothing major, so I will pre-emptively approve this PR, and let you decide on what to do about each suggestion.

File: ethology/annotations/json_schemas/schemas/VIA_schema.json [hunk 4]
```diff
@@ -0,0 +1,88 @@
+{
+    "$schema": "https://json-schema.org/draft/2020-12/schema",
```
[the remark this answers is not in the document]
Comment (sfmig) [reply ethology/annotations/json_schemas/schemas/VIA_schema.json:2]: Not really tbh, it is currently the latest metaschema. I figured since I am providing these two schemas I should specify a metaschema that I know works with them. If I leave it blank, it will use the most recent JSON metaschema, which I guess in the future it _could_ break backwards compatibility and not work for these?

File: ethology/annotations/json_schemas/utils.py [hunk 5]
```diff
@@ -0,0 +1,159 @@
...
+    )
+
+    # Raise error if there are missing keys in the schema
+    if missing_keys:
+        raise ValueError(
+            f"Required key(s) {sorted(missing_keys)} not found "
+            "in schema. Note that "
+            "a key may not be found correctly if the schema keywords "
+            "(such as 'properties', 'type' or 'items') are not spelt "
+            "correctly."
```
[the remark this answers is not in the document]
Comment (sfmig) [reply ethology/annotations/json_schemas/utils.py:73]: yeah the thing is that keys and keywords there refer to different things.

So some keys in the input JSON file are required, and they may not be found in the schema either (1) because they are not there, or (2) because the retrieval fails due to some of the schema keywords being misspelt. I thought the second reason may be a bit tricky to catch, but at the moment that is not the case anymore. If that happens, it is covered by a test that has a clear error message, so I am removing this bit.

Thanks for the heads up!

File: ethology/annotations/json_schemas/utils.py [hunk 6]
```diff
@@ -0,0 +1,159 @@
...
+    """Check if the required keys are present in the input dictionary."""
+    missing_keys = set(list_required_keys) - set(data.keys())
+    if missing_keys:
+        raise ValueError(
+            f"Required key(s) {sorted(missing_keys)} not "
+            f"found{additional_message}."
+        )
+
+
+def _extract_properties_keys(schema: dict, parent_key="") -> list:
```
[the remark this answers is not in the document]
Comment (sfmig) [reply ethology/annotations/json_schemas/utils.py:81]: fantastic, this is much more readable!

thanks Niko 🌟
